src/lein_monolith/task/fingerprint.clj

```diff
@@ -0,0 +1,352 @@
+  [project]
+  (->> (all-paths project)
+       (mapcat list-all-files)
+       (map hash-file)
+       (aggregate-hashes)))
+
+
+(defn- hash-dependencies
+  [project]
+  (-> (:dependencies project)
+      (pr-str)
+      (->multihash)))
+
+
+(declare hash-inputs)
+
+
+(defn- hash-upstream-projects
+  [project dep-map subprojects cache]
+  (->> (dep-map (dep/project-name project))
+       (keep subprojects)
+       (map #(::final (hash-inputs % dep-map subprojects cache)))
+       (aggregate-hashes)))
+
+
+(defn- cache-result!
+  [cache project m]
+  (get (swap! cache assoc (dep/project-name project) m) (dep/project-name project)))
+
+
+(defn- hash-inputs
+  "Hashes each of a project's inputs, and returns a map containing each individual
+  result, so it's easier to explain what aspect of a project caused its overall
+  fingerprint to change.
+
+  Returns a map of `{::xyz <mhash>}`
+
+  Keeps a cache of hashes computed so far, for efficiency."
+  [project dep-map subprojects cache]
+  (or (@cache (dep/project-name project))
+      (let [prints
+            {::sources (hash-sources project)
+             ::deps (hash-dependencies project)
+             ::upstream (hash-upstream-projects project dep-map subprojects cache)}
+
+            prints
+            (assoc prints
+                   ::final (aggregate-hashes (vals prints))
+                   ::time (System/currentTimeMillis))]
+        (cache-result! cache project prints))))
+
+
+;; ## Storing fingerprints
+
+;; The .lein-monolith-fingerprints file at the metaproject root stores the
+;; detailed fingerprint map for each project and marker type.
+
+(comment
+  ;; Example .lein-monolith-fingerprints
+  {:build {foo/bar {::sources "multihash abcde"
+                    ,,,
+                    ::final "multihash vwxyz"}
+           ,,,}
+   ,,,})
+
+
+(defn- fingerprints-file
+  ^File
+  [monolith]
+  (jio/file (:root monolith) ".lein-monolith-fingerprints"))
+
+
+(defn- read-fingerprints-file
+  [monolith]
+  (let [f (fingerprints-file monolith)]
+    (when (.exists f)
+      (edn/read-string
+        {:readers {'data/hash mhash/decode}}
+        (slurp f)))))
+
+
+(defn- write-fingerprints-file!
+  [monolith fingerprints]
+  (let [f (fingerprints-file monolith)]
+    (spit f (puget/pprint-str
+              fingerprints
+              {:print-handlers
+               {multihash.core.Multihash
+                (puget/tagged-handler 'data/hash mhash/base58)}}))))
+
+
+(let [lock (Object.)]
+  (defn update-fingerprints-file!
+    [monolith f & args]
+    (locking lock
+      (write-fingerprints-file!
+        monolith
+        (apply f (read-fingerprints-file monolith) args)))))
+
+
+;; ## Generating and comparing fingerprints
+
+(defn context
+  "Create a stateful context to use for fingerprinting operations."
+  [monolith subprojects]
+  (let [dep-map (dep/dependency-map subprojects)
+        initial (read-fingerprints-file monolith)
+        cache (atom {})]
+    {:monolith monolith
```

Comment:
This seems backwards - do you actually need the entire monolith project here? From a quick scan it looks like you might just need the root or fingerprint file location.